File: CodeForces/BlueRedPermutation.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.lang.Exception;
import java.util.ArrayList;
import java.util.Collections;

public class BlueRedPermutation{
    static class FastReader{
        BufferedReader br;
        StringTokenizer st;
        public FastReader(){
            br=new BufferedReader(new InputStreamReader(System.in));
        }
        String next(){
            while(st==null||!st.hasMoreElements()){
                try{
                    st=new StringTokenizer(br.readLine());
                }catch(IOException e){
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }
        int nextInt(){return Integer.parseInt(next());}
        long nextLong(){return Long.parseLong(next());}
        double nextDouble(){return Double.parseDouble(next());}
        String nextLine(){
            String str="";
            try{
                if(st.hasMoreTokens()) str=st.nextToken("\n");
                else str=br.readLine();
            }catch(IOException e){
                e.printStackTrace();
            }
            return str;
        }
    }
    static FastReader sc=new FastReader();
    static void solve(){
        int n=sc.nextInt();
        int[] v=new int[n];
        for(int i=0;i<n;i++){
            v[i]=sc.nextInt();
        }
        String colors=sc.next();
        ArrayList<Integer> blue=new ArrayList<Integer>(n);
        ArrayList<Integer> red=new ArrayList<Integer>(n);
        for(int i=0;i<n;i++){
            if(colors.charAt(i)=='B') blue.add(v[i]);
            else red.add(v[i]);
        }
        Collections.sort(blue);
        Collections.sort(red);
        int szb=blue.size();
        int szr=red.size();
        for(int i=0;i<szb;i++){
            if(blue.get(i)<i+1){
                System.out.println("NO");
                return;
            }
        }
        for(int i=0;i<szr;i++){
            if(red.get(i)>i+szb+1){
                System.out.println("NO");
                return;
            }
        }
        System.out.println("YES");
    }
    public static void main(String args[])throws java.lang.Exception{
        int t=sc.nextInt();
        while(t-->0){
            solve();
        }
    }
}

File: CodeForces/BlueRedPermutation.cpp
#include<iostream>
#include<string>
#include<vector>
#include<algorithm>

void solve(){
    int n;
    std::cin>>n;
    std::vector<int> v(n);
    for(int i=0;i<n;i++){
        std::cin>>v[i];
    }
    std::string colors;
    std::cin>>colors;
    std::vector<int> blue;
    std::vector<int> red;
    blue.reserve(n);
    red.reserve(n);
    for(int i=0;i<n;i++){
        if(colors[i]=='B') blue.push_back(v[i]);
        else red.push_back(v[i]);
    }
    std::sort(blue.begin(),blue.end());
    std::sort(red.begin(),red.end());
    int szb=blue.size();
    int szr=red.size();
    for(int i=0;i<szb;i++){
        if(blue[i]<i+1){
            std::cout<<"NO\n";
            return;
        }
    }
    for(int i=0;i<szr;i++){
        if(red[i]>i+szb+1){
            std::cout<<"NO\n";
            return;
        }
    }
    std::cout<<"YES\n";
}

int main(){
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);
    std::cout.tie(NULL);
    int t;
    std::cin>>t;
    while(t--){
        solve();
    }
    return 0;
}
